File: LeftBicepCurlStrategy.py
from ExerciseAnalysisStrategy import IExerciseAnalysisStrategy
from AngleCalculator import AngleCalculator
import mediapipe as mp

class LeftBicepCurlStrategy(IExerciseAnalysisStrategy):

    mp_pose = mp.solutions.pose.PoseLandmark

    def __init__(self, landmarks):
        
        self.hip_l = landmarks[self.mp_pose.LEFT_HIP.value]

        self.shoulder_l = landmarks[self.mp_pose.LEFT_SHOULDER.value]

        self.wrist_l = landmarks[self.mp_pose.LEFT_WRIST.value]
        
        self.elbow_l = landmarks[self.mp_pose.LEFT_ELBOW.value]

        self.l_elbow_angle = AngleCalculator.calculate_angle_from_landmarks(landmarks,
                                                                            self.mp_pose.LEFT_SHOULDER,
                                                                            self.mp_pose.LEFT_ELBOW,
                                                                            self.mp_pose.LEFT_WRIST)
        
        self.l_shoulder_angle = AngleCalculator.calculate_angle_from_landmarks(landmarks,
                                                                            self.mp_pose.LEFT_ELBOW,
                                                                            self.mp_pose.LEFT_SHOULDER,
                                                                            self.mp_pose.LEFT_HIP)
        
        self.l_shoulder_wrist = AngleCalculator.calculate_horizontal_distance(self.shoulder_l, self.wrist_l, 3)

    def is_correct(self):
        bicep_guide = ""

        if 20 < self.l_elbow_angle < 60 and 10 < self.l_shoulder_angle < 20 and 0.1 >= self.l_shoulder_wrist > 0 :
            bicep_guide = 'Keep Going!'

        else:
            #Conditon 1
            if 20 > self.l_elbow_angle:
                bicep_guide += 'Lower your wrist\n'
            if 60 < self.l_elbow_angle:
                bicep_guide += 'Raise your wrist\n'    
            #Condition 2
            if 0.1 < self.l_shoulder_wrist:
                bicep_guide += 'Move your wrist to the Right\n'
            if 0.05 > self.l_shoulder_wrist:
                bicep_guide += 'Move your wrist to the Left\n'
            # #Condition 3
            if 10 > self.l_shoulder_angle:
                bicep_guide += 'Raise your elbow\n'
            if 20 < self.l_shoulder_angle:
                bicep_guide += 'Lower your elbow\n'    

        print(self.l_shoulder_wrist)
        return bicep_guide

File: js/strategies/leftBicepCurlStrategy.js
import { IExerciseAnalysisStrategy } from './exerciseAnalysisStrategy.js';
import { AngleCalculator } from '../angleCalculator.js';
import { POSE } from '../landmarks.js';

export class LeftBicepCurlStrategy extends IExerciseAnalysisStrategy {
    constructor(landmarks) {
        super(landmarks);

        this.hip_l = landmarks[POSE.LEFT_HIP];
        this.shoulder_l = landmarks[POSE.LEFT_SHOULDER];
        this.wrist_l = landmarks[POSE.LEFT_WRIST];
        this.elbow_l = landmarks[POSE.LEFT_ELBOW];

        this.l_elbow_angle = AngleCalculator.calculateLandmarksAngle(landmarks,
            POSE.LEFT_SHOULDER, POSE.LEFT_ELBOW, POSE.LEFT_WRIST);

        this.l_shoulder_angle = AngleCalculator.calculateLandmarksAngle(landmarks,
            POSE.LEFT_ELBOW, POSE.LEFT_SHOULDER, POSE.LEFT_HIP);

        this.l_shoulder_wrist = AngleCalculator.calculateHorizontalDistance(this.shoulder_l, this.wrist_l, 3);
    }

    correctForm() {
        let bicepGuide = '';

        if (20 < this.l_elbow_angle && this.l_elbow_angle < 60 &&
            0.1 >= this.l_shoulder_wrist && this.l_shoulder_wrist > -0.1) {
            bicepGuide = 'Keep Going!';
        } else {
            // Condition 1
            if (20 > this.l_elbow_angle) {
                bicepGuide += 'Lower your wrist\n';
            }
            if (60 < this.l_elbow_angle) {
                bicepGuide += 'Raise your wrist\n';
            }
            // Condition 2
            if (0.1 < this.l_shoulder_wrist) {
                bicepGuide += 'Move your wrist to the Right\n';
            }
            if (-0.1 > this.l_shoulder_wrist) {
                bicepGuide += 'Move your wrist to the Left\n';
            }
        }

        return bicepGuide;
    }
}
